gcd/src/main/java/org/spine3/server/storage/datastore/tenant/Namespace.java

```diff
@@ -0,0 +1,222 @@
+     */
+    static Namespace of(TenantId id) {
+        checkNotNull(id);
+
+        final String value;
+        final TenantIdConverter converter;
+        final TenantId.KindCase kindCase = id.getKindCase();
+        switch (kindCase) {
+            case DOMAIN:
+                value = id.getDomain()
+                          .getValue();
+                converter = TenantIdConverter.DOMAIN;
+                break;
+            case EMAIL:
+                value = id.getEmail()
+                          .getValue();
+                converter = TenantIdConverter.EMAIL;
+                break;
+            case VALUE:
+                value = id.getValue();
+                converter = TenantIdConverter.VALUE;
+                break;
+            case KIND_NOT_SET:
+            default:
+                throw new IllegalArgumentException(format("Tenant ID is not set. Kind of TenantId is %s.",
+                                                          kindCase.toString()));
+        }
+        return new Namespace(value, converter);
+    }
+
+    private static String escapeIllegalCharacters(String candidateNamespace) {
+        final String result = AT_SYMBOL_PATTERN.matcher(candidateNamespace)
+                                               .replaceAll(Matcher.quoteReplacement(AT_SYMBOL_REPLACEMENT));
+        return result;
+    }
+
+    /**
+     * @return a string value of this {@code Namespace}
+     */
+    public String getValue() {
+        return value;
+    }
+
+    /**
+     * Converts this object to a {@link TenantId}.
+     *
+     * <p>If current instance was created with {@link Namespace#of(TenantId)}, then the result will
+     * be {@code equal} to that {@link TenantId}.
+     *
+     * <p>If current instance was created with {@link Namespace#of(String)}, then the result will
+     * be equivalent to the result of
+     * <code>
+     * <pre>
+     *         TenantId.newBuilder()
+     *                 .setValue(namespace.getValue())
+     *                 .build();
+     *     </pre>
+     * </code>
+     *
+     * @return a {@link TenantId} represented by this {@code Namespace}
+     */
+    TenantId toTenantId() {
+        final TenantId tenantId = tenantIdConverter.apply(this);
+        return tenantId;
+    }
+
+    @Override
+    public String toString() {
+        return getValue();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Namespace namespace = (Namespace) o;
+        return Objects.equal(getValue(), namespace.getValue());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(getValue());
+    }
+
+    /**
+     * The converter of the {@code Namespace} into a {@link TenantId} of the specific type..
+     */
+    private enum TenantIdConverter implements Function<Namespace, TenantId> {
```

Comment:
It's `TenantIdFunction`, not a converter, as it doesn't implement or extend `Converter`.

You could have a `Converter`, if you merge your `TenantId` -> `Namespace` and `Namespace` -> `TenantId` conversions in an instance of [`Converter`](https://google.github.io/guava/releases/20.0/api/docs/com/google/common/base/Converter.html).